bpf/bpf.go

```diff
@@ -1024,6 +1025,59 @@ func (b *BPFLib) getMapArgs(ifName string) ([]string, error) {
 	return mapArgs, nil
 }
 
+func (b *BPFLib) getCalicoBPFProgIds() ([]int, error) {
+	// Get a list of ids of BPF programs that loaded by Calico.
+	// Programs are identified by 'name', inherited from func name.
+	// for example, in bpf/bpf-apache/filter.c, a function name is 'prefilter'.
+	progName := "prefilter"
+
+	prog := "bpftool"
+	args := []string {
+		"--json",
+		"--pretty",
+		"prog",
+		"show",
+	}
+	printCommand(prog, args...)
+	output, err := exec.Command(prog, args...).CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get list of BPF programs: %s\n%s", err, output)
+	}
+	var p []progInfo
+	err = json.Unmarshal(output, &p)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse json output: %v\n%s", err, output)
+	}
+
+	var id = []int{}
+	for _, v := range p {
+		if v.Name == progName {
+			id = append(id, v.Id)
+		}
+	}
+
+	return id, nil
+}
+
+func (b *BPFLib) canRemoveXDP(ifName string, id []int) (bool, error) {
+	i, err := b.GetXDPID(ifName)
+	if err != nil {
+		// iproute2 error or Empty
+		if i == -1 {
+			return false, err
+		} else {
```

Comment:
And this one.

Reply:
-1: Failed to get interface information
-2: Failed to get XDP prog ID, maybe just XDP program is not loaded yet.

-1 is clearly a situation we don't want, but in case of -2, we have to proceed RemoveXDP().

Reply:
It will probably be better to encode the type of error in the Error object - but I'd like to make sure I understand the overall picture first.